cartographer/mapping_3d/local_trajectory_builder.cc

```diff
@@ -152,46 +177,62 @@ LocalTrajectoryBuilder::AddAccumulatedRangeData(
   const sensor::PointCloud low_resolution_point_cloud_in_tracking =
       low_resolution_adaptive_voxel_filter.Filter(filtered_range_data.returns);
   ceres_scan_matcher_->Match(
-      matching_submap->local_pose().inverse() * scan_matcher_pose_estimate_,
+      matching_submap->local_pose().inverse() * pose_prediction,
       initial_ceres_pose,
       {{&filtered_point_cloud_in_tracking,
         &matching_submap->high_resolution_hybrid_grid()},
        {&low_resolution_point_cloud_in_tracking,
         &matching_submap->low_resolution_hybrid_grid()}},
       &pose_observation_in_submap, &summary);
-  const transform::Rigid3d pose_observation =
-      matching_submap->local_pose() * pose_observation_in_submap;
-  pose_tracker_->AddPoseObservation(
-      time, pose_observation,
-      options_.kalman_local_trajectory_builder_options()
-              .scan_matcher_variance() *
-          kalman_filter::PoseCovariance::Identity());
+  pose_estimate_ = matching_submap->local_pose() * pose_observation_in_submap;
 
-  scan_matcher_pose_estimate_ = pose_tracker_->GetPoseEstimateMean(time);
+  odometry_correction_ = transform::Rigid3d::Identity();
+  if (!odometry_state_tracker_.empty()) {
+    // We add an odometry state, so that the correction from the scan matching
+    // is not removed by the next odometry data we get.
+    odometry_state_tracker_.AddOdometryState(
+        {time, odometry_state_tracker_.newest().odometer_pose,
+         odometry_state_tracker_.newest().state_pose *
+             odometry_prediction.inverse() * pose_estimate_});
+  }
+
+  // Improve the velocity estimate.
+  if (last_scan_match_time_ > common::Time::min() &&
+      time > last_scan_match_time_) {
+    const double delta_t = common::ToSeconds(time - last_scan_match_time_);
+    velocity_estimate_ +=
+        (pose_estimate_.translation() - model_prediction.translation()) /
```

Comment:
Wouldn't we be better off updating velocity based on the most recent two estimates after refinement by the scan matcher? This appears to be comparing pose_estimate_, as last updated by scan matching, with the previous pose_estimate_, as last updated by prediction.